### ISSUE TITLE
enum.IntFlag.name is typed as str|None

### ISSUE THREAD

**Issue by adaamz** [python/mypy#16177]
Hello,
I would like to ask/report bug whether it is possible to have `(enum.IntFlag)().name` as type None or if it is incorrectly typed as `str|None`.

**Bug Report**
Looks like `(enum.IntFlag)().name` is typed as `str|None` but I think it should be just `str`.

**To Reproduce**
```python
import enum


class ListType(enum.IntFlag):
    ABC = enum.auto()
    DEF = enum.auto()
    GHI = enum.auto()
    
    def get_names(self) -> list[str]:
        to_return = []

        for e in ListType:
            reveal_type(e)   # note: Revealed type is "__main__.ListType"
            reveal_type(e.name)  # note: Revealed type is "Union[builtins.str, None]"
            to_return.append(e.name)

        return to_return  # error: Incompatible return value type (got "list[str | None]", expected "list[str]")  [return-value]

```
https://onlinegdb.com/Rsv6ryZzf
https://mypy-play.net/?mypy=latest&python=3.11&gist=61af435c15e57ce5221abada2221008a

**Expected Behavior**
I would expect that `(enum.IntFlag)().name` is always string.
```
Success: no issues found in 1 source file
```

**Actual Behavior**
```
main.py:19: error: Incompatible return value type (got "list[str | None]", expected "list[str]")  [return-value]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.5.1
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.9, 3.10, 3.11

**Comment by JelleZijlstra** [comment]
I believe this is correct specifically for Flag. We have an explicit override for it: https://github.com/python/typeshed/blob/9e79a100b3a9f1ae931b46cea56dc3ef99122ae5/stdlib/enum.pyi#L235 Flags can have a None name because they can have values (combinations of basic flags) without an associated name.

In your case, you're iterating over the predefined members of the enum, which will always have a name, but I don't think there's a way to express that in the type system.

**Comment by adaamz** [comment]
@JelleZijlstra thank you for quick response. How can I create such enum without a name? I'm just curious how and why to do it, as I have never seen it 😉

**Comment by adaamz** [comment]
Thanks for examples, looks like enum is pretty strange in Python 🙂 

I will just use cast to string around `e.name` on this single place.

**Comment by JelleZijlstra** [comment]
I would recommend using `assert e.name is not None`, for what it's worth; that communicates the same intent and is also safe at runtime.
